{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install biopython\n",
    "from Bio import SeqIO \n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pysam\n",
    "import time\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lecture des données\n",
    "Ouverture et lecture des fichiers contenant les lectures et le génome de réference."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "reads = list(SeqIO.parse('single_Pfal_dat.fq','fastq'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "Genome = list(SeqIO.parse('data','fasta'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Definition des fonctions utilisées"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def complementInverse(w):\n",
    "    wCI = []\n",
    "    for i in range(len(w)):\n",
    "        li = w[-i-1]\n",
    "        if li not in \"ATCG\":\n",
    "            raise Exception(\"Letter not in alphabet\")\n",
    "        lCI = 'A'*int(li=='T') + 'T'*int(li=='A') + 'C'*int(li=='G') + 'G'*int(li=='C')\n",
    "        wCI.append(lCI)\n",
    "    return ''.join(wCI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def circularShift(T,k):\n",
    "    if k<0:\n",
    "        raise Exception(\"k must be positive\")\n",
    "    return T[k:]+T[0:k]\n",
    "\n",
    "def BWT(S):\n",
    "    if S[len(S)-1]!='$':\n",
    "        S += '$'\n",
    "        \n",
    "    circShifts = []\n",
    "    for k in range(len(S)):\n",
    "        circShifts.append(circularShift(S,k))\n",
    "    circShifts.sort()\n",
    "    \n",
    "    bwt = []\n",
    "    for shift in circShifts:\n",
    "        bwt.append(shift[len(S)-1])\n",
    "        \n",
    "    return bwt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def suffix_list(T): # O(n log n)\n",
    "    \"\"\"\n",
    "    Compute the suffix list\n",
    "    \n",
    "    Args:\n",
    "        T (str): string\n",
    "    \n",
    "    Return:\n",
    "        list of strings: suffix list\n",
    "    \"\"\"\n",
    "    suffix_list = [T[i:] for i in range(len(T))] # O(n)\n",
    "    sorted(suffix_list,reverse=True) # O(n log n)\n",
    "    return suffix_list\n",
    "\n",
    "def suffix_table(T): # O(n log n)\n",
    "    \"\"\"\n",
    "    Compute the suffix table\n",
    "    \n",
    "    Args:\n",
    "        T (str): string\n",
    "    \n",
    "    Return:\n",
    "        list of tuples (suffix,location): suffix table\n",
    "    \"\"\"\n",
    "    suffix_list = [T[i:] for i in range(len(T))] \n",
    "    suffix_table = sorted((e,i) for i,e in enumerate(suffix_list))\n",
    "    return suffix_table\n",
    "\n",
    "def BWT_suffix_table(T,end_of_string=\"$\"): # O(n log n)\n",
    "    \"\"\"\n",
    "    Compute the BWT from the suffix table\n",
    "    \n",
    "    Args:\n",
    "        T (str): string\n",
    "        end_of_string (char): end of string character to append\n",
    "    \n",
    "    Return:\n",
    "        bwt (str): BWT\n",
    "    \"\"\"\n",
    "    if T[-1]!=end_of_string :\n",
    "        T += end_of_string\n",
    "    ST = suffix_table(T) # O(n log n)\n",
    "    bwt = \"\"\n",
    "    index = []\n",
    "    for s,i in ST: # O(n)\n",
    "        bwt += T[i-1]\n",
    "        index.append(i-1)\n",
    "    return bwt, index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter \n",
    "\n",
    "def occurrence_indexer(S):\n",
    "    \"\"\"\n",
    "    Number of past occurrences of each char in S\n",
    "    \n",
    "    Args:\n",
    "        S (str): string\n",
    "        \n",
    "    Return:\n",
    "        table of int : number of past occurrences of the char\n",
    "    \"\"\"\n",
    "    K = []\n",
    "    last_index = {}\n",
    "    for s in S:\n",
    "        if s not in last_index: \n",
    "            last_index[s] = 0\n",
    "        K.append(last_index[s])\n",
    "        last_index[s] += 1\n",
    "    return(K)\n",
    "\n",
    "def last2first(counts,k,X):\n",
    "    \"\"\"\n",
    "    k + nbr of occurrences of letters < X \n",
    "    \n",
    "    Args:\n",
    "        counts (Counter): nbr of each char in the text\n",
    "        k : int\n",
    "        X : char\n",
    "        \n",
    "    Return:\n",
    "        int : k + nbr of occurrences of letters < X \n",
    "    \"\"\"\n",
    "    a = [counts[char] for char in counts if char < X]\n",
    "    return k + sum(a)\n",
    "    \n",
    "def efficient_inverse_BWT(bwt,end_of_string=\"$\"):\n",
    "    \"\"\"\n",
    "    Inverse the BWT\n",
    "    \n",
    "    Args:\n",
    "        bwt (str): bwt of a string T\n",
    "        last_character (char): which is the end of string character?\n",
    "    \n",
    "    Return:\n",
    "        T (str): BWT^{-1} of bwt\n",
    "    \"\"\"    \n",
    "    K = occurrence_indexer(bwt)\n",
    "    counts = Counter(bwt)\n",
    "    X = bwt[0]\n",
    "    k = K[0]\n",
    "    S = end_of_string\n",
    "    while X != end_of_string:\n",
    "        S = X+S\n",
    "        j = last2first(counts,k,X)\n",
    "        X = bwt[j]\n",
    "        k = K[j]\n",
    "    return(S)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "  def get_first_occurrence(L,X):\n",
    "    for i,l in enumerate(L):\n",
    "        if l == X:\n",
    "            return(i)\n",
    "        \n",
    "def pattern_matching_BWT(T,pattern):\n",
    "    \"\"\"\n",
    "    Search a patter in a String using the BWT\n",
    "    \n",
    "    Args:\n",
    "        S (str): string\n",
    "        pattern (str): pattern\n",
    "    \n",
    "    Return:\n",
    "        bool: true if the pattern is in the string    \n",
    "    \"\"\"\n",
    "    L,_ = BWT_suffix_table(T)\n",
    "    K = occurrence_indexer(L)\n",
    "    counts = Counter(L)\n",
    "    e = 0\n",
    "    f = len(L)\n",
    "    i = len(pattern) - 1\n",
    "    while e < f and i >= 0:\n",
    "        X = pattern[i]\n",
    "        first_occurence_in_L_ef = get_first_occurrence(L[e:f],X)\n",
    "        if first_occurence_in_L_ef is None:\n",
    "            return False\n",
    "        else:\n",
    "            r = first_occurence_in_L_ef+e\n",
    "        #print(first_occurence_in_L_ef,L[e:f],r,X,L[r])\n",
    "        \n",
    "        last_occurence_in_L_ef = get_first_occurrence(L[e:f][::-1],X)\n",
    "        if last_occurence_in_L_ef is None:\n",
    "            return False\n",
    "        else:\n",
    "            s = f-last_occurence_in_L_ef-1\n",
    "        #print(last_occurence_in_L_ef,L[e:f],s,X,L[s])\n",
    "        \n",
    "        e = last2first(counts,K[r],X)\n",
    "        f = last2first(counts,K[s],X)+1\n",
    "        i -= 1\n",
    "        #print(r,s,e,f)\n",
    "    return(i<0)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rankBWT(bwt):\n",
    "    \"\"\"\n",
    "    Number of past occurrences of each char in S\n",
    "    \n",
    "    Args:\n",
    "        S (str): string\n",
    "        \n",
    "    Return:\n",
    "        table of int : number of past occurrences of the char\n",
    "    \"\"\"\n",
    "    K = []\n",
    "    tot = {}\n",
    "    for s in bwt:\n",
    "        if s not in tot: \n",
    "            tot[s] = 0\n",
    "        K.append(tot[s])\n",
    "        tot[s] += 1\n",
    "    return(K,tot)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def invBWT(bwt,A):\n",
    "    A = sorted(A)\n",
    "    #count = {a:0 for a in A}\n",
    "    count = {}\n",
    "    K = []\n",
    "    for i in range(len(bwt)):\n",
    "        if bwt[i] in count:\n",
    "            count[bwt[i]] += 1\n",
    "        else:\n",
    "            count[bwt[i]] = 1\n",
    "        K.append(count[bwt[i]])\n",
    "    X = bwt[0]\n",
    "    k = K[1]\n",
    "    S = '$'\n",
    "    while X != '$':\n",
    "        S = X+S\n",
    "        j = k\n",
    "        i = 0\n",
    "        while A[i] != X:\n",
    "            j += count[A[i]]\n",
    "            i += 1\n",
    "        X = bwt[j]\n",
    "        k = K[j]\n",
    "    return S"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def firstColMap(tots):\n",
    "    \"\"\" Return a map from characters to the range of cells in the first\n",
    "    column containing the character. \n",
    "    char : (first index, last index+1)\"\"\"\n",
    "    first = {}\n",
    "    totc = 0\n",
    "    for c, count in sorted(tots.items()):\n",
    "        first[c] = (totc, totc + count)\n",
    "        totc += count\n",
    "    return first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def searching(bwt, SAT, W):\n",
    "\n",
    "    # Not case sensitive, make everything uppercase\n",
    "    bwt = bwt.upper()\n",
    "    W = W.upper()\n",
    "    \n",
    "    # Perdorm a BWT on text T and get the indexes corresponding to the Suffix Array of T\n",
    "    #bwt, SAT = BWT_suffix_table(T)\n",
    "    # LF mapping\n",
    "    # F = sorted(bwt) # no need to store it\n",
    "    rank, totCount = rankBWT(bwt)\n",
    "    Fmap = firstColMap(totCount)\n",
    "    del totCount\n",
    "\n",
    "    \n",
    "    #print('\\nbwt :', sys.getsizeof(bwt),\n",
    "    #'\\nSAT :', sys.getsizeof(SAT),\n",
    "    #'\\nrank :', sys.getsizeof(rank),\n",
    "    #'\\nFmap :', sys.getsizeof(Fmap),)\n",
    "\n",
    "    letter = W[-1] # last letter of the word W\n",
    "    if letter not in Fmap:\n",
    "        return -1, \"Word not in text\"\n",
    "    fi, li = Fmap[letter] # find occurences of letter in F\n",
    "    \n",
    "    for i in range(2,len(W)+1):\n",
    "        letter = W[-i] # search W by going backward through it\n",
    "        if letter not in Fmap:\n",
    "            return -1, \"Word not in text\"\n",
    "\n",
    "        j = fi\n",
    "        foundRankf = False\n",
    "        while j < li:\n",
    "            if bwt[j] == letter:\n",
    "                if not foundRankf:\n",
    "                    rankf = rank[j] # get the rank of the first occurence of letter (followed by W suffix)\n",
    "                    foundRankf = True\n",
    "                rankl = rank[j] # get the rank of the last occurence of letter (followed by W suffix)\n",
    "            j += 1\n",
    "        \n",
    "        if not foundRankf:\n",
    "            return -1, \"Word not in text\"\n",
    "        \n",
    "        fi, li = Fmap[letter] # find occurences of letter in F\n",
    "        fi += rankf # find first occurence of letter in F (followed by W suffix)\n",
    "        li = fi + rankl-rankf +1 # find last occurence of letter in F (followed by W suffix)\n",
    "\n",
    "    return [x+1 for x in SAT[fi:li]]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def searchWord(cutSeq, n, k, words):\n",
    "\n",
    "    bwtSeq = []\n",
    "    satSeq = []\n",
    "    for s in cutSeq:\n",
    "        bwt, sat = BWT_suffix_table(s)\n",
    "        bwtSeq.append(bwt)\n",
    "        satSeq.append(sat)\n",
    "\n",
    "    foundRead = []\n",
    "    for i in range(len(cutSeq)):\n",
    "        res = searching(bwtSeq[i], satSeq[i], words)\n",
    "        if res[0] != -1 :\n",
    "            resi = [x+i*(n-k) for x in res]\n",
    "            foundRead += resi\n",
    "\n",
    "    return np.unique(foundRead)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def GenomeBWT(genome):\n",
    "    bwtGen = []\n",
    "    satGen = []\n",
    "    \n",
    "    wholeGenomeCutted = []\n",
    "    n = int(1e4)\n",
    "    overlap = 100\n",
    "    for i in range(len(Genome)):\n",
    "        chromosome = Genome[i].seq.upper()\n",
    "        cutChr = cutSequence(chromosome, n, overlap)\n",
    "        wholeGenomeCutted.append(cutChr)\n",
    "        \n",
    "    for c in range(len(genome)):\n",
    "        print('chr',c)\n",
    "        chromosome = wholeGenomeCutted[c]\n",
    "        bwtChr = []\n",
    "        satChr = []\n",
    "        for seq in chromosome:\n",
    "            bwt, sat = BWT_suffix_table(seq)\n",
    "            bwtChr.append(bwt)\n",
    "            satChr.append(sat)\n",
    "        bwtGen.append(bwtChr)\n",
    "        satGen.append(satChr)\n",
    "    return (bwtGen,satGen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def searchChr(bwtChr,satChr, word):\n",
    "    n = int(1e4)\n",
    "    overlap = 100\n",
    "\n",
    "    foundRead = []\n",
    "    for i in range(len(bwtChr)):\n",
    "        res = searching(bwtChr[i], satChr[i], word)\n",
    "        if res[0] != -1 :\n",
    "            resi = [x+i*(n-overlap) for x in res]\n",
    "            foundRead += resi\n",
    "\n",
    "    return np.unique(foundRead)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def searchGen(bwtGen,satGen, word):\n",
    "   \n",
    "    foundGen = {}\n",
    "    for i in range(len(bwtGen)):\n",
    "        foundChr = searchChr(bwtGen[i],satGen[i], word)\n",
    "        foundGen[i] = foundChr\n",
    "        \n",
    "    return foundGen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getKmers(read, k):\n",
    "    read = str(read)\n",
    "    kmers = []\n",
    "    iF = 0\n",
    "    iL = iF + k\n",
    "\n",
    "    Continuer = True\n",
    "\n",
    "    while Continuer:\n",
    "        if iL >= len(read):\n",
    "            Continuer = False\n",
    "            iL = len(read)\n",
    "        kmers.append(read[iF:iL])\n",
    "        iF +=1\n",
    "        iL = iF + k\n",
    "        if Continuer and iF >= len(read):\n",
    "            Continuer = False\n",
    "            \n",
    "    return kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def SWmatrix2strings(S,T,D,F):\n",
    "    \"\"\"\n",
    "    Smith-Waterman matrix to aligned strings\n",
    "\n",
    "    Args:\n",
    "        S (str): first string\n",
    "        T (str): second string\n",
    "        D (np.array or list of lists): Decision matrix\n",
    "\n",
    "    Returns:\n",
    "        S_aligned (str): first string aligned\n",
    "        T_aligned (str): second string aligned\n",
    "        \n",
    "    \"\"\"    \n",
    "    S_WITH_GAP = -1\n",
    "    T_WITH_GAP = 1\n",
    "    S_T_ALIGN = 0\n",
    "    i,j = np.unravel_index(F.argmax(),F.shape)\n",
    "    end = (i,j)\n",
    "    S_aligned = \"\"\n",
    "    T_aligned = \"\"\n",
    "    while i > 0 and j > 0:\n",
    "        #print(i,j,T[j-1],S[i-1],D[i,j])\n",
    "        #print(S_aligned,T_aligned)\n",
    "        #print(S[i-1], S_aligned)\n",
    "        if D[i,j] == S_WITH_GAP:\n",
    "            S_aligned = S[i-1]+S_aligned\n",
    "            T_aligned = \"_\"+T_aligned\n",
    "            #print(S_aligned,T_aligned)\n",
    "            i -= 1\n",
    "        elif D[i,j] == T_WITH_GAP:\n",
    "            T_aligned = T[j-1]+ T_aligned\n",
    "            S_aligned = \"_\"+S_aligned\n",
    "            #print(S_aligned,T_aligned)\n",
    "            j -= 1\n",
    "        elif D[i,j] == S_T_ALIGN:\n",
    "            S_aligned = S[i-1] + S_aligned\n",
    "            T_aligned = T[j-1] + T_aligned\n",
    "            #print(S_aligned,T_aligned)\n",
    "            j -= 1\n",
    "            i -= 1\n",
    "        #elif D[i,j] == NO_ALIGN:\n",
    "        else:\n",
    "            break\n",
    "    start = (i,j)\n",
    "    return(S_aligned,T_aligned,start,end)\n",
    "\n",
    "def Smith_Waterman(S,T,return_matrix=False):\n",
    "    \"\"\"\n",
    "    Needleman-Wunch \n",
    "\n",
    "    Args:\n",
    "        S (str): first string\n",
    "        T (str): second string\n",
    "        return_matrix (bool): If true return matrix, else return the aligned strings\n",
    "\n",
    "    Returns:\n",
    "        F,D (np.array, np.array): If return_matrix is true F is the score matrix and D the decision matrix\n",
    "        S_aligned, T_aligned (str,str): If return_matrix is false first and second strings aligned\n",
    "        \n",
    "    \"\"\"\n",
    "    T_WITH_GAP = 1\n",
    "    S_WITH_GAP = -1\n",
    "    S_T_ALIGN = 0\n",
    "    NO_ALIGN = 2\n",
    "    decision = [T_WITH_GAP,S_WITH_GAP,S_T_ALIGN,NO_ALIGN]\n",
    "    F = np.zeros((len(S)+1,len(T)+1))\n",
    "    D = np.zeros((len(S)+1,len(T)+1))\n",
    "    D[0,1:] = T_WITH_GAP \n",
    "    D[1:,0] = S_WITH_GAP \n",
    "    for i in range(1,len(S)+1):\n",
    "        for j in range(1,len(T)+1):\n",
    "            comparison = int(S[i-1]==T[j-1])*2-1\n",
    "            options = [F[i,j-1] - 1, F[i-1,j] - 1, F[i-1,j-1] + comparison, 0]\n",
    "            F[i,j] = options[0]\n",
    "            D[i,j] = T_WITH_GAP\n",
    "            for o,val in enumerate(options):\n",
    "                if val > F[i,j]:\n",
    "                    F[i,j] = val\n",
    "                    D[i,j] = decision[o]\n",
    "    if return_matrix:\n",
    "        return F,D\n",
    "    else:\n",
    "        return SWmatrix2strings(S,T,D,F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cutSequence(seqToCut, n, k):\n",
    "    cutSeq = []\n",
    "    iF = 0\n",
    "    iL = iF + n\n",
    "\n",
    "    Continuer = True\n",
    "\n",
    "    while Continuer:\n",
    "        if iL >= len(seqToCut):\n",
    "            Continuer = False\n",
    "            iL = len(seqToCut)\n",
    "        cutSeq.append(seqToCut[iF:iL])\n",
    "        iF += n - k\n",
    "        iL = iF + n -1\n",
    "        if Continuer and len(seqToCut) - iF <= k :\n",
    "            Continuer = False\n",
    "            \n",
    "    return cutSeq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "def MappingKmers(Read,ReadSeq,bwtGen,satGen, iCh,k, CompInv=False):\n",
    "    \n",
    "    KmersRead = getKmers(ReadSeq, k)\n",
    "    uniqueKmers = np.unique(KmersRead)\n",
    "    kmerDict = {k : [] for k in uniqueKmers}\n",
    "    for i in range(len(KmersRead)):\n",
    "        kmerDict[KmersRead[i]].append(i)\n",
    "    \n",
    "    ChrKmersPos = []\n",
    "    #ChrKmersFound = {}\n",
    "    ChrKmersFreq = {}\n",
    "    for w in uniqueKmers:\n",
    "        foundChr = searchChr(bwtGen[iCh],satGen[iCh], w)\n",
    "        #ChrKmersFound[w] = foundChr\n",
    "        ChrKmersFreq[w] = len(foundChr)\n",
    "        for i in foundChr:\n",
    "            ChrKmersPos.append((w,i))\n",
    "            \n",
    "            \n",
    "    # Filter out very frequent k-mers\n",
    "    ChrKmersPosFilt = []\n",
    "    threshold = np.percentile(list(ChrKmersFreq.values()),75)\n",
    "    for t in ChrKmersPos:\n",
    "        if ChrKmersFreq[t[0]] < threshold:\n",
    "            ChrKmersPosFilt.append(t)\n",
    "\n",
    "    # Sort k-mers along chromosome\n",
    "    ChrKmersPosFiltdf = pd.DataFrame(ChrKmersPosFilt, columns=['K-mer','i']).sort_values('i')\n",
    "    SortedKmers = list(ChrKmersPosFiltdf.loc[:,'K-mer'])\n",
    "\n",
    "    # Chromosom-read k-mers correspondance  \n",
    "    corrRead = [] # indices des k-mers de la lecture mis en correspondance avec les k-mers du génome\n",
    "    i = -1\n",
    "    for k in SortedKmers: # Pour chaque k-mer dans le génome\n",
    "        i += 1\n",
    "        listI = kmerDict[k] # liste des positions possibles dans la lecture\n",
    "        # Si qu'une position ou que début de la mise en correspondance \n",
    "        # ou qu'on a déjà une fin de lecture mis en correspondance juste avant\n",
    "        if len(listI) == 1 or len(corrRead) == 0 or listI[-1]-corrRead[-1]<=0:\n",
    "            corrRead.append(listI[0]) # on met en correspondance k du génome avec la première occurence de k dans le read\n",
    "        else:\n",
    "            # différence entre les posible occurences de k dans read \n",
    "            # avec l'index de la dernière correspondance faite\n",
    "            dist = np.array(listI)-corrRead[-1]\n",
    "            j = 0\n",
    "            # on cherche l'occurence pouvant étant après l'index de la dernière correspondance faite\n",
    "            # et la plus proche de celle-ci\n",
    "            while dist[j]<=0:\n",
    "                j+=1\n",
    "\n",
    "            corrRead.append(listI[j])    \n",
    "    \n",
    "    x = corrRead\n",
    "    y = list(ChrKmersPosFiltdf.loc[:,'i'])\n",
    "    \n",
    "    return (x,y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "def findRead(x,y,Read,ReadSeq,Genome,iCh,CompInv=False):\n",
    "    \n",
    "    yRange = (min(y),max(y))\n",
    "    ySpan = yRange[1]-yRange[0]\n",
    "    epsilon = len(ReadSeq)*2\n",
    "\n",
    "    mi = yRange[0]\n",
    "    ma = mi+epsilon\n",
    "    i = 0\n",
    "    yMov = [0]*ySpan\n",
    "    for j in range(len(y)):\n",
    "        if y[j] < ma:\n",
    "            yMov[i] += 1\n",
    "        else:\n",
    "            lastyMov = yMov[i]\n",
    "            while y[j] >= ma:\n",
    "                i += 1\n",
    "                mi += 1\n",
    "                ma += 1\n",
    "                yMov[i] = yMov[i-1]\n",
    "                jMi = j-yMov[i-1]\n",
    "                while y[jMi] < mi:\n",
    "                    yMov[i] -= 1 \n",
    "                    jMi += 1\n",
    "            yMov[i] += 1\n",
    "     \n",
    "    # Supposons que la lecture ne peux pas être présente plus d'une fois dans un chromosome.\n",
    "    iStart = yMov.index(max(yMov))+yRange[0]\n",
    "    iEnd = iStart + epsilon\n",
    "    iStart -= int(len(ReadSeq)/2)\n",
    "    iEnd += int(len(ReadSeq)/2)\n",
    "    #print(iStart)\n",
    "    #print(iEnd)\n",
    "    \n",
    "    S = str(Genome[iCh].seq)[iStart:iEnd].upper()\n",
    "    T = ReadSeq\n",
    "    F,D = Smith_Waterman(S,T,True)\n",
    "    seqChr,seqRead,istart,iend = SWmatrix2strings(S,T,D,F)\n",
    "    \n",
    "    mappedTo = (istart[0]+iStart,iend[0]+iStart)\n",
    "    Fscore = np.amax(F)\n",
    "    #print(Fscore)\n",
    "    \n",
    "    if CompInv:\n",
    "        flag = 16\n",
    "    else:\n",
    "        flag = 0\n",
    "    this = np.reshape([flag,iCh,mappedTo[0],mappedTo[1],Fscore],(1, 5))\n",
    "    thisMapping = pd.DataFrame(this,index = [Read.id],columns=['Flag','chr', 'iStart', 'iEnd', 'Fscore'])\n",
    "    \n",
    "    return thisMapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def MappingOneRead(Genome,reads,iRead,k,bwtGen,satGen,CompInv=False,temp=False):\n",
    "    # on va chercher la bonne lecture en prenant la complémentaire inverse si CompInv = True\n",
    "    Read = reads[iRead]\n",
    "    if CompInv:\n",
    "        ReadSeq = complementInverse(str(Read.seq).upper())\n",
    "    else:\n",
    "        ReadSeq = str(Read.seq).upper()\n",
    "        \n",
    "    # on récupère la liste des k-mers de la lecture    \n",
    "    KmersRead = getKmers(ReadSeq, k)\n",
    "    uniqueKmers = np.unique(KmersRead)\n",
    "    kmerDict = {k : [] for k in uniqueKmers}\n",
    "    for i in range(len(KmersRead)):\n",
    "        kmerDict[KmersRead[i]].append(i)\n",
    "        \n",
    "    # initalisation des variables de stockage de temps de calcul\n",
    "    timeMapKmers = []\n",
    "    timeFindRead = []    \n",
    "    # initalisation de la variable de stockage de l'alignement  \n",
    "    storeMapping = pd.DataFrame(columns=['Flag','chr', 'iStart', 'iEnd', 'Fscore'])\n",
    "    # pour chaque chromosome\n",
    "    for iCh in range(len(Genome)):\n",
    "        #print(iCh)\n",
    "        t0 = time.perf_counter()\n",
    "        x,y = MappingKmers(Read,ReadSeq,bwtGen,satGen, iCh,k,CompInv) # où sont les k-mers dans le génome ?\n",
    "        t1 = time.perf_counter()\n",
    "        A = findRead(x,y,Read,ReadSeq,Genome,iCh,CompInv) # il y a-t-il des zones du génome pouvant être une lecture ?\n",
    "        t2 = time.perf_counter()\n",
    "        print(A)\n",
    "        timeMapKmers.append(t1-t0)\n",
    "        timeFindRead.append(t2-t1)\n",
    "        # on va considérer qu'il n'existe qu'une bonne occurence de la lecture sur le génome\n",
    "        if int(A.loc[:,'Fscore']) >= 90:\n",
    "            storeMapping = storeMapping.append(A)\n",
    "            break\n",
    "            \n",
    "    if temp:\n",
    "        return (storeMapping,timeMapKmers,timeFindRead)\n",
    "    return storeMapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Description de l'algorithme\n",
    "\n",
    "1. Une transformée de Burrows-Wheeler (BWT) est appliquée au génome de référence. \n",
    "2. La lecture est décomposée en k-mers.\n",
    "3. Les différents k-mers présents dans la lecture sont recherchées et positionnées dans le génome de référence en utilisant la BWT et la table de suffixes des chromosomes composants le génome.\n",
    "4. La fréquence de chaque k-mers dans le génome est calculée et les k-mers très fréquents (suppérieur au $75^e$ percentile) sont exclus de l'étude.\n",
    "5. On met en correspondance les position des k-mers dans le génome (y) avec les position des k-mers dans la lecture (x)\n",
    "6. Afin de détecter la présence de la lecture dans le génome, on cherche une zone de y où l'on va avoir la présence d'une majorité des positions de k-mers sur x. Pour cela on récupère le nombre de points tels que $y\\in[a,b]$ avec $[a,b]$ une fenêtre glissante le long de y. On suppose qu'une lecture peut être présente dans le génome lorque l'on a un nombre de points élevé dans la fenêtre glissante.\n",
    "7. On réalise un alignement de cette zone "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Création de la variable de stockage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "storeMapping = pd.DataFrame(columns=['Flag','chr', 'iStart', 'iEnd', 'Fscore'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Préprocessing du génome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On applique une transformée de Burrows-Wheeler (BWT) à tout le génome en le séparant en plusieurs sous séquences pour ne pas surcharger la mémoire vive :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr 0\n",
      "chr 1\n",
      "chr 2\n",
      "chr 3\n",
      "chr 4\n",
      "chr 5\n",
      "chr 6\n",
      "chr 7\n",
      "chr 8\n",
      "chr 9\n",
      "chr 10\n",
      "chr 11\n",
      "chr 12\n",
      "chr 13\n",
      "chr 14\n"
     ]
    }
   ],
   "source": [
    "bwtGen, satGen = GenomeBWT(Genome)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cette transformée est couteuse et n'est donc effectuée qu'une fois pour tout le génome."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Recherche d'une lecture dans le génome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On prend une des lectures (ici la première)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "iRead = 0\n",
    "k = 10\n",
    "CompInv = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'satGenCompInv' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-65-0c76e3700cf3>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mMappingOneRead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mGenome\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mreads\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0miRead\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mbwtGen\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msatGenCompInv\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mtrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'satGenCompInv' is not defined"
     ]
    }
   ],
   "source": [
    "storeMapping,timeMapKmers,timeFindRead = MappingOneRead(Genome,reads,iRead,k,bwtGen,satGen,CompInv,True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'Time taken to find read (s)')"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEKCAYAAAAfGVI8AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAezUlEQVR4nO3de5wcZZ3v8c+XIcLIbVaILkkIQcRRkEt0QBFcEJQED0K8rMJ6BV8gZ8HL6mZN9oiCrkfdiLf1EgMbAVflhRJiECRwUEDBSxIHMoQ4mOUimbAG1CGII4TwO39UNXQmPd01k6nunqrv+/XqV3c9VdX1e7pn6tdVT9XzKCIwM7Py2qHVAZiZWWs5EZiZlZwTgZlZyTkRmJmVnBOBmVnJORGYmZVcbolA0mJJGyXdOcJ8SfqypHWSVkt6aV6xmJnZyPI8IrgEmF1n/onAAenjLODrOcZiZmYjyC0RRMQtwB/rLHIKcFkkfgF0Sdo7r3jMzKy2HVu47anAA1XT69OyB+uttNdee8WMGTNyDMvMrHhWrVr1cERMrjWvlYlANcpq9nch6SyS00dMnz6dlStX5hmXmVnhSLp/pHmtvGpoPbBP1fQ0YEOtBSNiUUT0RETP5Mk1E5qZmY1RKxPBMuCd6dVDrwAeiYi6p4XMzGz85XZqSNJ3gWOBvSStBz4OTAKIiIXAtcDrgHXAX4DT84rFzMxGllsiiIjTGswP4Jy8tm9mZtn4zmIzs5Jr5VVDZmaWwdLeARYs72fD4BBTujqZO6ubOTOnjtv7OxGYmbWxpb0DzF/Sx9DmLQAMDA4xf0kfwLglA58aMjNrYwuW9z+dBCqGNm9hwfL+cduGE4GZWRvbMDg0qvKxcCIwM2tjU7o6R1U+Fk4EZmZtbO6sbjondWxV1jmpg7mzusdtG24sNjNrY5UGYV81ZGZWYnNmTh3XHf9wPjVkZlZyTgRmZiXnRGBmVnJOBGZmJedEYGZWck4EZmYl50RgZlZyTgRmZiXnRGBmVnJOBGZmJedEYGZWck4EZmYl50RgZlZyTgRmZiXnRGBmVnJOBGZmJeeBaczMGljaO5DrCGGt5kRgZlbH0t4B5i/pY2jzFgAGBoeYv6QPoDDJwKeGzMzqWLC8/+kkUDG0eQsLlve3KKLx50RgZlbHhsGhUZVPRE4EZmZ1TOnqHFX5ROREYGZWx9xZ3XRO6tiqrHNSB3NndbcoovHnxmIzszoqDcK+asjMrMTmzJxaqB3/cD41ZGZWck4EZmYl50RgZlZyTgRmZiWXayKQNFtSv6R1kubVmL+HpKsl3SFpjaTT84zHzMy2lVsikNQBfBU4ETgQOE3SgcMWOwe4KyIOBY4FLpT0rLxiMjOzbTW8fFTSzsBJwKuAKcAQcCdwTUSsqbPqEcC6iLgnfZ/LgVOAu6qWCWA3SQJ2Bf4IPDmGepiZ2RjVTQSSzgdeD9wE/BLYCOwMvBD4TJokPhwRq2usPhV4oGp6PfDyYct8BVgGbAB2A94aEU+NuhZmZjZmjY4IVkTE+SPM+7yk5wLTR5ivGmUxbHoWcDtwHLA/cIOkn0bEpq3eSDoLOAtg+vSRNmdmZmNRt40gIq4ZXiZpB0m7p/M3RsTKEVZfD+xTNT2N5Jd/tdOBJZFYB9wLvKhGHIsioicieiZPnlwvZDMzG6VMjcWSviNpd0m7kJzj75c0t8FqK4ADJO2XNgCfSnIaqNrvgOPTbTwP6AbuGU0FzMxs+2S9aujA9HTNHOBaktNB76i3QkQ8CZwLLAfWAldExBpJZ0s6O13sk8ArJfUBNwIfiYiHx1APMzMbo6ydzk2SNIkkEXwlIjZLGn6+fxsRcS1J4qguW1j1egNwwijiNTOzcZb1iOAbwH3ALsAtkvYFNtVdw8zMJoRMiSAivhwRUyPidRERJOf2X51vaGZm1gx1E4Gkt0vaZpn0Kp8nJe0v6ej8wjMzs7w1aiPYE+iVtApYBTxEckPZC4BjgIeBbfoQMjOziaNuIoiIL0n6CskNX0cBh5B0MbEWeEdE/C7/EM3MLE8NrxqKiC3ADenDzMwKxuMRmJmVnBOBmVnJORGYmZVco26oP1RvfkR8fnzDMTOzZmvUWLxb+twNHM4znca9Hrglr6DMzKx5Gl0+egGApOuBl0bEo+n0+cD3co/OzMxyl7WNYDrwRNX0E8CMcY/GzMyaLmvvo98CfiXpKpJRxt4AXJZbVGZm1jSZEkFEfErSdUClX6HTI6I3v7DMzKxZsh4REBGrJD1A0tcQkqa7iwmz8lnaO8CC5f1sGBxiSlcnc2d1M2fm1FaHZdshUyKQdDJwITAF2EjSZvAb4KD8QjOzdrO0d4D5S/oY2rwFgIHBIeYv6QNwMpjAsjYWfxJ4BXB3ROwHvAa4NbeozKwtLVje/3QSqBjavIUFy/tbFJGNh6yJYHNE/AHYQdIOEfET4LAc4zKzNrRhcGhU5TYxZG0jGJS0K/BT4NuSNgJP5heWmbWjKV2dDNTY6U/p6mxBNDZesh4RnAL8BfggcB3w3yR3F5tZicyd1U3npI6tyjondTB3VneLIrLxkPXy0cfSAesPiIhLJT0b6Gi0npkVS6VB2FcNFUvWq4bOBM4CngPsD0wFFgLH5xeambWjOTOnesdfMFlPDZ1DMlTlJoCI+C3w3LyCMjOz5smaCB6PiKf7GpK0I0lXE2ZmNsFlTQQ3S/pXoFPSa0l6Hr06v7DMzKxZsiaCjwAPAX3Ae4FrgY/mFZSZmTVPw8ZiSTsAqyPiJcBF+YdkZmbN1PCIICKeAu6QNL0J8ZiZWZNlvbN4b2CNpF8Bj1UKI+LkXKIyM7OmyZoILsg1CjMza5msdxbfnHcgZmbWGpkHpjGziceDyFgWTgRmBeVBZCyrrPcRmNkE40FkLKu6RwSS+qjTlUREHDLuEZnZuPAgMpZVo1NDJ6XP56TP30qf30YyPoGZtSkPImNZ1T01FBH3R8T9wFER8S8R0Zc+5gGzmhOimY2FB5GxrLK2Eewi6ejKhKRXArs0WknSbEn9ktZJmjfCMsdKul3SGkm+TNVsnMyZOZVPv/FgpnZ1ImBqVyeffuPBbii2bWS9aug9wGJJe6TTg8AZ9VaQ1AF8FXgtsB5YIWlZRNxVtUwX8DVgdkT8TpLHODAbRx5ExrLIekPZKuBQSbsDiohHMqx2BLAuIu4BkHQ5ydjHd1Ut8w/Akoj4XbqdjaMJ3szMtl/WoSp3At4EzAB2lARARHyizmpTgQeqptcDLx+2zAuBSZJuAnYDvhQRl9XY/lkkQ2Uyfbr7vjMzG09ZTw39AHgEWAU8nnEd1SgbfinqjsDLSMY+7gR+LukXEXH3VitFLAIWAfT09HhkNDOzcZQ1EUyLiNmjfO/1wD7V7wFsqLHMwxHxGPCYpFuAQ4G7MTOzpsh61dBtkg4e5XuvAA6QtJ+kZwGnAsuGLfMD4FWSdpT0bJJTR2tHuR0zM9sOWY8IjgbeLeleklNDAqLencUR8aSkc4HlQAewOCLWSDo7nb8wItZKug5YDTwFXBwRd25HfczMbJQU0fiUu6R9a5WnN5s1VU9PT6xcubLZmzUzm9AkrYqInlrzGvU1tHtEbAIezSUyMzNruUanhr5D0t/QKpIrfqqvBArg+TnFZWZmTdIoEXwmfX5xRPw172DMzKz5Gl019KX0+ba8AzEzs9ZodESwWdI3gWmSvjx8ZkS8P5+wzMysWbKMR/Aa4DiSdgIzMyuYuokgIh4GLpe0NiLuaFJMZmbWRJnuLHYSMDMrLg9eb2ZWck4EZmYllykRSNpD0hckrUwfF1aNVmZmZhNY1iOCxcAm4C3pYxPwzbyCMjOz5sna++j+EfGmqukLJN2eR0BmE93S3gEWLO9nw+AQU7o6mTur2+MGW1vLmgiGJB0dET8DkHQUMJRfWGatN5Yd+tLeAeYv6WNo8xYABgaHmL+kD8DJwNpW1kRwNnBZVbvAn4B35ROSWeuNdYe+YHn/0+tUDG3ewoLl/U4E1raythFsiohDgUOAQyJiJu6a2gqs3g69ng2DtQ+URyo3awdZE8GVABGxKR2fAOD7+YRk1npj3aFP6eocVblZO2g0MM2LgIOAPSS9sWrW7sDOeQZm1gwjtQNM6epkoMZOv9EOfe6s7q1OKQF0Tupg7qzucY/dbLw0aiPoJul4rgt4fVX5o8CZeQVl1gz12gHGukOvtAP4qiGbSLKOWXxkRPy8CfE05DGLbXtUHwHsILGlxt//1K5Obp13nC8DtUIZ85jFFe2SBMy2x/AjgFpJAJ5pB5gzc6p3/FYK7mvISqPWlUC1uGHXysaJwEojyyWcbti1Msp0akjSTsCbgBnV60TEJ/IJy2z87dE5icGhzduUS0DgdgArrax3Fv8AeIRkuMrH8wvHLD9S7fKuzkn0fuyE5gZj1kayJoJpETE710jMcjb4l22PBuqVm5VF1kRwm6SDI6Iv12jMxsF43yRmVnRZG4uPBlZJ6pe0WlKfpNV5BmY2FpVLRAcGhwieuUlsae8Ac2d10zmpY6vl3Thslv2I4MRcozAbJ/U6i7t13nFPL+ObxMyekfWGsvslHQ0cEBHflDQZ2DXf0MxGr1Fncb5JzGxbWccs/jjwEWB+WjQJ+K+8gjIbK/f+aTZ6WdsI3gCcDDwGEBEbgN3yCspsrNwOYDZ6WdsInoiIkBQAknbJMSazMXPvn2ajlzURXCHpG0CXpDOBM4CL8wvLbOzcDmA2Olkbiz8n6bXAJpIxCj4G3JJnYGZm1hxZ+xpaHBFnADek07sC1wLH5xibmZk1QdbG4gFJXweQ9DfA9fiqITOzQsiUCCLiPGCTpIUkSeDCiPhmo/UkzU7vRl4naV6d5Q6XtEXSmzNHbqWwtHeAoz7zY/abdw1HfebHLO0daHVIZoXTaPD66gHrfwWclz6HpDdGxJI663YAXwVeC6wHVkhaFhF31Vjus8DysVXBiqremMJuDDYbP42OCF5f9TgJ6CW5mawyXc8RwLqIuCcingAuB06psdz7gCuBjaOI20qgXncRZjZ+6h4RRMTp2/HeU4EHqqbXAy+vXkDSVJKb1Y4DDh/pjSSdBZwFMH369O0IySaSRt1FmNn4yHrV0M7Ae4CDgJ0r5emVRCOuVqNs+GjhXwQ+EhFbNNKoIcl2FgGLAHp6emqPOG6F426jzZoj61VD3wL+FpgF3AxMAx5tsM56YJ+q6WnAhmHL9ACXS7oPeDPwNUlzMsZkBefuIsyaI+udxS+IiL+XdEpEXCrpOzRu3F0BHCBpP2AAOBX4h+oFImK/ymtJlwA/jIilmaO3CWGkgWIacXcRZs2RNRFUxvIblPQS4H9IBrIfUUQ8KelckoTRASyOiDWSzk7nLxxbyDaRbO+VP+4uwix/WRPBovRGso8Cy0jGIjiv0UoRcS3JHcjVZTUTQES8O2MsNoHUu/LHO3iz9pA1EdwYEX8i6V/o+QDpKR+zunzlj1n7y9pYfGWNsu+PZyBWTB4oxqz9Nbqz+EUkl4zuMewu492puozUbCRzZ3Vv1UYAvvLHrN00OjXUTXIHcRfJ3cQVjwJn5hWUFYev/DFrf4pofH+WpCMj4udNiKehnp6eWLlyZavDMDObUCStioieWvOy9j7aFknAzMzGX9bGYjMzKygnAjOzksuUCCQ9T9J/SvpROn2gpPfkG5qZmTVD1iOCS0i6ipiSTt8NfDCPgGxi8QhiZhNf1kSwV0RcATwFST9CwJb6q1jRVfoRGhgcInimHyEnA7OJJWsieEzSnqTjCUh6BfBIblHZhOARxMyKIWtfQx8i6Wxuf0m3ApNJxg+wEnM/QmbFkCkRRMSvJR1DcqexgP6I2NxgNSs4jyBmVgxZrxrqAF4HHA+cALxP0ofyDMzan0cQMyuGrKeGrgb+CvSRNhibuR8hs2LImgimRcQhuUZiE5JHEDOb+LJeNfQjSSfkGomZmbVE1iOCXwBXSdqBZPxiARERu+cWmZmZNUXWRHAhcCTQF1n6rTYzswkj66mh3wJ3OgmYmRVP1iOCB4Gb0k7nHq8URsTnc4nKzMyaJmsiuDd9PCt9mJlZQWS9s/iCvAMxM7PWqJsIJH0lIs6VdDVph3PVIuLk3CIzM7OmaHRE8E7gXOBzTYjFttPS3gHf5Wtmo9YoEfw3QETc3IRYbDtUxgaodAtdGRsAcDIws7oaJYLJ9TqX81VD7aPe2ABOBGZWT6NE0AHsSnInsbUxjw1gZmPVKBE8GBGfaEoktl08NoCZjVWjO4t9JDBBeGwAMxurRkcExzclCttuHhvAzMaqbiKIiD82KxDbfh4bwMzGImunc2ZmVlBOBGZmJedEYGZWck4EZmYll2sikDRbUr+kdZLm1Zj/Nkmr08dtkg7NMx4zM9tWbolAUgfwVeBE4EDgNEkHDlvsXuCYiDgE+CSwKK94zMystjyPCI4A1kXEPRHxBHA5cEr1AhFxW0T8KZ38BTAtx3jMzKyGPBPBVOCBqun1adlI3gP8qNYMSWdJWilp5UMPPTSOIZqZWZ6JoFb3FNsMbgMg6dUkieAjteZHxKKI6ImInsmTJ49jiGZmlnXM4rFYD+xTNT0N2DB8IUmHABcDJ0bEH3KMx8zMasjziGAFcICk/SQ9CzgVWFa9gKTpwBLgHRFxd46xmJnZCHI7IoiIJyWdCywnGddgcUSskXR2On8h8DFgT+BrkgCejIievGIyM7NtKaLmafu21dPTEytXrmx1GGZmE4qkVSP90PadxWZmJedEYGZWck4EZmYl50RgZlZyTgRmZiXnRGBmVnJOBGZmJedEYGZWck4EZmYl50RgZlZyTgRmZiXnRGBmVnJOBGZmJedEYGZWcnmOUNY2lvYOsGB5PxsGh5jS1cncWd3MmVlv+GQzs/IofCJY2jvA/CV9DG3eAsDA4BDzl/QBOBmYmVGCU0MLlvc/nQQqhjZvYcHy/hZFZGbWXgqfCDYMDo2q3MysbAqfCKZ0dY6q3MysbAqfCObO6qZzUsdWZZ2TOpg7q7tFEZmZtZfCNxZXGoR91ZCZWW2FTwSQJAPv+M3Maiv8qSEzM6vPicDMrOScCMzMSs6JwMys5JwIzMxKThHR6hhGRdJDwP2jWGUv4OGcwmknrmexuJ7F0g713DciJteaMeESwWhJWhkRPa2OI2+uZ7G4nsXS7vX0qSEzs5JzIjAzK7kyJIJFrQ6gSVzPYnE9i6Wt61n4NgIzM6uvDEcEZmZWR6ETgaTZkvolrZM0r9XxjBdJiyVtlHRnVdlzJN0g6bfp89+0MsbtJWkfST+RtFbSGkkfSMuLVs+dJf1K0h1pPS9IywtVzwpJHZJ6Jf0wnS5cPSXdJ6lP0u2SVqZlbV3PwiYCSR3AV4ETgQOB0yQd2Nqoxs0lwOxhZfOAGyPiAODGdHoiexL4cES8GHgFcE76/RWtno8Dx0XEocBhwGxJr6B49az4ALC2arqo9Xx1RBxWdcloW9ezsIkAOAJYFxH3RMQTwOXAKS2OaVxExC3AH4cVnwJcmr6+FJjT1KDGWUQ8GBG/Tl8/SrLzmErx6hkR8ed0clL6CApWTwBJ04D/BVxcVVy4eo6gretZ5EQwFXiganp9WlZUz4uIByHZiQLPbXE840bSDGAm8EsKWM/0dMntwEbghogoZD2BLwL/AjxVVVbEegZwvaRVks5Ky9q6nkUemEY1ynyJ1AQjaVfgSuCDEbFJqvW1TmwRsQU4TFIXcJWkl7Q6pvEm6SRgY0SsknRsq+PJ2VERsUHSc4EbJP2m1QE1UuQjgvXAPlXT04ANLYqlGX4vaW+A9Hlji+PZbpImkSSBb0fEkrS4cPWsiIhB4CaS9p+i1fMo4GRJ95Gcpj1O0n9RvHoSERvS543AVSSnqdu6nkVOBCuAAyTtJ+lZwKnAshbHlKdlwLvS1+8CftDCWLabkp/+/wmsjYjPV80qWj0np0cCSOoEXgP8hoLVMyLmR8S0iJhB8r/444h4OwWrp6RdJO1WeQ2cANxJm9ez0DeUSXodyXnJDmBxRHyqxSGNC0nfBY4l6dHw98DHgaXAFcB04HfA30fE8AblCUPS0cBPgT6eOaf8ryTtBEWq5yEkjYcdJD/MroiIT0jakwLVs1p6auifI+KkotVT0vNJjgIgOfX+nYj4VLvXs9CJwMzMGivyqSEzM8vAicDMrOScCMzMSs6JwMys5JwIzMxKzomgxSTtmfZSeLuk/5E0kL7+s6Sv5bC9OVk635N0iaQ3t2LbzSbpfEn/3Oo4miHtGXOvVsfRCpK+n17eOdL8z0k6rpkxtQsnghaLiD+kvRQeBiwEvpBO7xoR/5jDJueQ9MbaCq3cdtuQ1NZdu7R7fGMh6SCgIyLuqbPYf9BmvYI2ixNBm5J0bFWf7edLulTS9ekvujdK+ve0z/Pr0q4YkPQySTennV0tr9zSXvWerwROBhakRx37SzpT0oq0P/wrJT27RiyfTI8QdpA0N11+tZ7pO3+GknEDLlLSp/716V2yjbZ9mKRfpO91Va0+2tPtfl3J2AT3SDpGyXgMayVdUrXc1yWtVFWf/mn5fZI+q6TP/19JekGDz/1MST+qEf8MSb+RdLGkOyV9W9JrJN2qpI/5I9LldknjW6Gk3/1T0vJ3S/qepKtJOiTbW9It6Wdxp6RX1Yjl+PQ9+tL33KmqThdI+nU670U11u1If+H2pZ/v+6pmv2/4uunf2CJJ1wOXSdpX0o3pujdKmj7K7+O09P3vlPTZqpguScv6JP1TWl7z70DSTZK+kH5OayUdLmlJ+nn/W9W23p5+t7dL+oaSLuiHexvp3bwjxRER9wN7Svrben8jhRQRfrTJAzif5I5LSO4c/mFV+c9Iuig+FPgLcGI67yqSX9qTgNuAyWn5W0nuph6+jUuAN1dN71n1+t+A91UvB/w78A2STvxOIBl7VSQ/In4I/B0wg2T8gMPSda8A3p5h26uBY9LXnwC+OMI6l6fbPAXYBBycbn9V1Tafkz53kPTXc0g6fR/wf9LX76x8prU+d+Bckq4AdqqxTKWO1dteXBXX0nS5/1upO9AF3A3sArybpP+rSpwfroqrA9ht2PZ2Juk994Xp9GUkHe9V6lT5nv4RuLhGvP+bpJ+mHYd9PjXXTT+DVUBnOn018K709RlV9Wv4fQBTSO6enUxyd+2PSf5GX0bSu2olxq56fwfp9/jZ9PUHSPoK2xvYKf0s9wRenMY6KV3ua8A7a3weNwMHp69rxpG+vgh4U6v3Bc1++Ihg4vhRRGwm6XKhA7guLe8j2Ul1Ay8h6e3wduCjJB3tNfISST+V1Efyq+mgqnnnkfyTvDeS/5IT0kcv8GvgRcAB6bL3RsTt6etVaUwjkrRH+t43p0WXkiSVWq5Ot98H/D4i+iLiKWBN1XbeIunXaWwHsfUpqO9WPR85wjbeQTKI0Zsi4vERlrl32LZvrIqrEscJwLz0O7iJZIc+PZ13QzzTrcAK4HRJ55PsoB4dtq3udHt3p9PDP59KJ3wjfdavARZGxJMAsXV3BiOtuywihtLXRwLfSV9/Czi6arlG38fhwE0R8VC6/W+nsd8DPF/Sf0iaDWzK8HdQ6R+sD1gTyTgVj6fvtQ9wPMmOfUX6mR8P1GoH2Bt4KH29TRxVy20kSWSlUrhzgQX2OEBEPCVpc/qPCEk/PDuS/EJbExEj7ehGcgkwJyLukPRukiORihXAyyQ9J92RCPh0RHyj+g2UjBdQvfPcAmx1amU7Vd77qWHbeQrYUdJ+JL/oD4+IP6WnKHauWi5GeF3tTpJfs9OAeyXtQ/JLE5K2m+tqbLs6rsr/kkiSSX/1m0t6OfDY00FE3CLp70gGavmWpAURcVn1KiPEWVHZ9hZq/x+Lkes60rqP1Vj26ZBrrF/z+yA5ctr2DZLv5lBgFnAO8Bbgn+psM8u2BFwaEfMbvM8Q6d/ECHGckS63c7psqfiIoDj6gcmSjoSkC2clDWTDPQrsVjW9G/CgknaGtw1b9jrgM8A1SnpUXA6coWSMACRNVdLnelZPbzsiHgH+VHVu/B0kh+9jsTvJTuwRSc8j+WVf7a1Vzz8f4T16gfcCyyRNiYgHIm3Ej4iFo4hlOck5eAFImllrIUn7kvTPfxFJL6svHbbIb4AZeqZNY7Sfz/XA2UobfiU9ZxTrQnKa8dT09dtITk1m9UvgGEl7pefrTwNuVnK10g4RcSXJ0eZLx+Hv4EbgzZW/QyVjA+9bY7m1wAvSZbaJo2q5F5L8KCgVHxEUREQ8oeRyzy+nh9s7kvS8umbYopcDF0l6P0kbwHkk/7j3kxx+VycJIuJ7aRJYBryO5HTBz9P93J+Bt5P8ssxi+LbfBSxU0kB9D3D6qCr9TIx3SOolqes9wK3DFtlJ0i9JfvicVud9fqbkMtJrJL02Ih4eQzifJPncV6fJ4D7gpBrLHQvMlbSZ5HN857BY/irpdOB76c58BcmRSVYXk+zUVqfbuAj4yijWfz+wWNJcklMqmb+biHhQ0nzgJyS/2K+NiB+kv8K/KanyA7TyK37MfwcRcZekj5I0wO8AbCb5lX//sEWvIfnM/x/JSIXbxJH+GHoBsDLr9ovCvY9aoSkZCKVnjDt1KwglV4H9hGT0sJo/XCS9geQo5bymBtcGfGrIzAovbQT/OPXHLd8RuLA5EbUXHxGYmZWcjwjMzErOicDMrOScCMzMSs6JwMys5JwIzMxKzonAzKzk/j+Q2H4OTtjItAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.scatter(timeMapKmers,timeFindRead)\n",
    "plt.xlabel(\"Time taken to map k-mers on chromosome (s)\")\n",
    "plt.ylabel(\"Time taken to find read (s)\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "L'étape prenant le plus de temps est de trouver toutes les occurences des k-mers sur les chromosomes pouvant aller jusqu'à 1min pour certain chromosomes. On a une relation $1/60$ entre le temps de mapping des k-mers et le positionnement d'un lecture à partir de ce mapping."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Idem pour le complément inverse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                    chr    iStart      iEnd  Fscore\n",
      "NC_004325.2-100000  0.0  131734.0  131834.0    98.0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(                    chr    iStart      iEnd  Fscore\n",
       " NC_004325.2-100000  0.0  131734.0  131834.0    98.0,\n",
       " [9.766329269999915],\n",
       " [0.25863098799982254])"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CompInv = True\n",
    "MappingOneRead(Genome,reads,iRead,k,bwtGen,satGen,CompInv)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comparaison à la vérité terrain\n",
    "Ouverture du fichier .bam :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "samfile = pysam.AlignmentFile('./single_Pfal_dat.bam', 'rb')\n",
    "align = samfile.fetch(until_eof=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cherchons les alignements de la lecture traitée"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for a in align:\n",
    "    if a.qname == Read.id:\n",
    "        print(a)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On retrouve un résultat similaire : avec la complementaire inverse de la lecture (flag = 16) alignée sur le chromosome 1 (#0) à la position $131735$.\n",
    "\n",
    "Pour rendre le résultat plus juste il faudrait revoir la détection de read à partir du mapping des k-mers;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mapping pour plusieurs lectures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "                    Flag  chr    iStart      iEnd  Fscore\n",
      "NC_004325.2-100000   0.0  0.0  104204.0  104297.0    37.0\n",
      "1\n",
      "                    Flag  chr    iStart      iEnd  Fscore\n",
      "NC_004325.2-100000   0.0  1.0  213397.0  213426.0    16.0\n",
      "2\n",
      "                    Flag  chr    iStart      iEnd  Fscore\n",
      "NC_004325.2-100000   0.0  2.0  665540.0  665613.0    35.0\n",
      "3\n",
      "                    Flag  chr    iStart      iEnd  Fscore\n",
      "NC_004325.2-100000   0.0  3.0  135268.0  135334.0    22.0\n",
      "4\n",
      "                    Flag  chr     iStart       iEnd  Fscore\n",
      "NC_004325.2-100000   0.0  4.0  1066826.0  1066903.0    25.0\n",
      "5\n",
      "                    Flag  chr    iStart      iEnd  Fscore\n",
      "NC_004325.2-100000   0.0  5.0  552090.0  552187.0    29.0\n",
      "6\n",
      "                    Flag  chr    iStart      iEnd  Fscore\n",
      "NC_004325.2-100000   0.0  6.0  855730.0  855826.0    34.0\n",
      "7\n",
      "                    Flag  chr    iStart      iEnd  Fscore\n",
      "NC_004325.2-100000   0.0  7.0  741449.0  741530.0    33.0\n",
      "8\n",
      "                    Flag  chr    iStart      iEnd  Fscore\n",
      "NC_004325.2-100000   0.0  8.0  827560.0  827647.0    38.0\n",
      "9\n",
      "                    Flag  chr    iStart      iEnd  Fscore\n",
      "NC_004325.2-100000   0.0  9.0  812632.0  812730.0    37.0\n",
      "10\n",
      "                    Flag   chr    iStart      iEnd  Fscore\n",
      "NC_004325.2-100000   0.0  10.0  724582.0  724671.0    32.0\n",
      "11\n",
      "                    Flag   chr     iStart       iEnd  Fscore\n",
      "NC_004325.2-100000   0.0  11.0  1406414.0  1406492.0    38.0\n",
      "12\n",
      "                    Flag   chr     iStart       iEnd  Fscore\n",
      "NC_004325.2-100000   0.0  12.0  1420693.0  1420796.0    30.0\n",
      "13\n",
      "                    Flag   chr     iStart       iEnd  Fscore\n",
      "NC_004325.2-100000   0.0  13.0  2252488.0  2252560.0    35.0\n",
      "14\n",
      "                    Flag   chr   iStart     iEnd  Fscore\n",
      "NC_004325.2-100000   0.0  14.0  29951.0  30040.0    29.0\n",
      "0\n",
      "                    Flag  chr    iStart      iEnd  Fscore\n",
      "NC_004325.2-100000  16.0  0.0  131734.0  131834.0    98.0\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "cannot convert the series to <class 'int'>",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-87-f9a02499bbc2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0mCompInv\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m     \u001b[0mB\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mMappingOneRead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mGenome\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mreads\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0miRead\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mbwtGen\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msatGen\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mCompInv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m     \u001b[0;32mif\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mA\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'Fscore'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m>\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mB\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'Fscore'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     11\u001b[0m         \u001b[0mstoreMapping\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstoreMapping\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mA\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.8/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    110\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    111\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mconverter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 112\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"cannot convert the series to {converter}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    113\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    114\u001b[0m     \u001b[0mwrapper\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34mf\"__{converter.__name__}__\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: cannot convert the series to <class 'int'>"
     ]
    }
   ],
   "source": [
    "TimeRead = []\n",
    "rangeiRead = range(5)\n",
    "for iRead in rangeiRead:\n",
    "    t0 = time.perf_counter()\n",
    "    k = 10\n",
    "    CompInv = False\n",
    "    A = MappingOneRead(Genome,reads,iRead,k,bwtGen,satGen,CompInv)\n",
    "    CompInv = True\n",
    "    B = MappingOneRead(Genome,reads,iRead,k,bwtGen,satGen,CompInv)\n",
    "    if int(A.loc[:,'Fscore'])>int(B.loc[:,'Fscore']):\n",
    "        storeMapping = storeMapping.append(A)\n",
    "    else:\n",
    "        storeMapping = storeMapping.append(B)\n",
    "    t1 = time.perf_counter()\n",
    "    TimeRead.append(t1-t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>iStart</th>\n",
       "      <th>iEnd</th>\n",
       "      <th>Fscore</th>\n",
       "      <th>Flag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NC_004325.2-100000</th>\n",
       "      <td>0.0</td>\n",
       "      <td>104204.0</td>\n",
       "      <td>104297.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_004325.2-100000</th>\n",
       "      <td>0.0</td>\n",
       "      <td>131734.0</td>\n",
       "      <td>131834.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_004325.2-99999</th>\n",
       "      <td>0.0</td>\n",
       "      <td>143900.0</td>\n",
       "      <td>144000.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_004325.2-99999</th>\n",
       "      <td>0.0</td>\n",
       "      <td>109528.0</td>\n",
       "      <td>109621.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_004325.2-99998</th>\n",
       "      <td>0.0</td>\n",
       "      <td>471737.0</td>\n",
       "      <td>471837.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_004325.2-99998</th>\n",
       "      <td>0.0</td>\n",
       "      <td>148724.0</td>\n",
       "      <td>148801.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_004325.2-99997</th>\n",
       "      <td>0.0</td>\n",
       "      <td>65426.0</td>\n",
       "      <td>65524.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_004325.2-99997</th>\n",
       "      <td>0.0</td>\n",
       "      <td>458641.0</td>\n",
       "      <td>458740.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_004325.2-99996</th>\n",
       "      <td>0.0</td>\n",
       "      <td>274963.0</td>\n",
       "      <td>275067.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_004325.2-99996</th>\n",
       "      <td>0.0</td>\n",
       "      <td>152677.0</td>\n",
       "      <td>152777.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_004325.2-100000</th>\n",
       "      <td>0.0</td>\n",
       "      <td>104204.0</td>\n",
       "      <td>104297.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_004325.2-100000</th>\n",
       "      <td>0.0</td>\n",
       "      <td>131734.0</td>\n",
       "      <td>131834.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_004325.2-99999</th>\n",
       "      <td>0.0</td>\n",
       "      <td>143900.0</td>\n",
       "      <td>144000.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_004325.2-99999</th>\n",
       "      <td>0.0</td>\n",
       "      <td>109528.0</td>\n",
       "      <td>109621.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_004325.2-100000</th>\n",
       "      <td>0.0</td>\n",
       "      <td>131734.0</td>\n",
       "      <td>131834.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_004325.2-99999</th>\n",
       "      <td>0.0</td>\n",
       "      <td>143900.0</td>\n",
       "      <td>144000.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    chr    iStart      iEnd  Fscore  Flag\n",
       "NC_004325.2-100000  0.0  104204.0  104297.0    37.0   0.0\n",
       "NC_004325.2-100000  0.0  131734.0  131834.0    98.0  16.0\n",
       "NC_004325.2-99999   0.0  143900.0  144000.0    98.0   0.0\n",
       "NC_004325.2-99999   0.0  109528.0  109621.0    31.0  16.0\n",
       "NC_004325.2-99998   0.0  471737.0  471837.0   100.0   0.0\n",
       "NC_004325.2-99998   0.0  148724.0  148801.0    26.0  16.0\n",
       "NC_004325.2-99997   0.0   65426.0   65524.0    98.0   0.0\n",
       "NC_004325.2-99997   0.0  458641.0  458740.0    49.0  16.0\n",
       "NC_004325.2-99996   0.0  274963.0  275067.0    26.0   0.0\n",
       "NC_004325.2-99996   0.0  152677.0  152777.0    98.0  16.0\n",
       "NC_004325.2-100000  0.0  104204.0  104297.0    37.0   0.0\n",
       "NC_004325.2-100000  0.0  131734.0  131834.0    98.0  16.0\n",
       "NC_004325.2-99999   0.0  143900.0  144000.0    98.0   0.0\n",
       "NC_004325.2-99999   0.0  109528.0  109621.0    31.0  16.0\n",
       "NC_004325.2-100000  0.0  131734.0  131834.0    98.0  16.0\n",
       "NC_004325.2-99999   0.0  143900.0  144000.0    98.0   0.0"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "storeMapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "align = samfile.fetch(until_eof=True)\n",
    "for a in align:\n",
    "    print(a.qname)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Le temps d'execution reste encore bien trop élevé pour pouvoir traiter toutes les lectures en un temps raisonnable. De plus, la méthode de détection d'une lecture dans les k-mers mappé sur le génome pourrait être retravaillé pour être plus précis et éventuellement pouvoir détecter plusieurs zones où une lecture pourrait être alignée. \n",
    "\n",
    "Un travail sur la sortie des données serait aussi à faire, pour avoir des données d'alignement bien formmatées et qui puissent être comparées au fichier bam fournis."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
